File: src/rider/main/kotlin/rider/RiderProjectOutputHost.kt
package me.fornever.avaloniarider.rider

import com.intellij.openapi.components.Service
import com.intellij.openapi.project.Project
import com.jetbrains.rd.ide.model.RdGetProjectOutputArgs
import com.jetbrains.rd.ide.model.RdProjectOutput
import com.jetbrains.rd.ide.model.riderProjectOutputModel
import com.jetbrains.rd.platform.util.withUiContext
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rider.projectView.solution
import me.fornever.avaloniarider.idea.concurrency.await
import java.nio.file.Path

@Service
class RiderProjectOutputHost(private val project: Project) {
    companion object {
        fun getInstance(project: Project): RiderProjectOutputHost =
            project.getService(RiderProjectOutputHost::class.java)
    }

    suspend fun getProjectOutput(lifetime: Lifetime, projectFilePath: Path): RdProjectOutput =
        withUiContext(lifetime) {
            val model = project.solution.riderProjectOutputModel
            model.getProjectOutput.start(lifetime, RdGetProjectOutputArgs(projectFilePath.toString())).await(lifetime)
        }
}

File: src/dotnet/AvaloniaRider.Plugin/RiderProjectOutputHost.cs
using System.Linq;
using JetBrains.Application.Components;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.ProjectModel;
using JetBrains.ProjectModel.ProjectsHost;
using JetBrains.ProjectModel.ProjectsHost.SolutionHost;
using JetBrains.Rd.Tasks;
using JetBrains.ReSharper.Host.Features;
using JetBrains.ReSharper.Host.Features.ProjectModel.TargetFrameworks;
using JetBrains.Rider.Model;
using JetBrains.Util;

namespace ReSharperPlugin.AvaloniaRider
{
    [SolutionComponent]
    public class RiderProjectOutputHost
    {
        private readonly ISolution _solution;
        private readonly ILogger _logger;

        public RiderProjectOutputHost(ISolution solution, ILogger logger)
        {
            _solution = solution;
            _logger = logger;

            var riderProjectOutputModel = solution.GetProtocolSolution().GetRiderProjectOutputModel();
            riderProjectOutputModel.GetProjectOutput.Set(GetProjectOutputModel);
        }

        private RdTask<RdProjectOutput> GetProjectOutputModel(Lifetime lifetime, RdGetProjectOutputArgs args)
        {
            var projectsHostContainer = _solution.ProjectsHostContainer();
            var solutionStructureContainer = projectsHostContainer.GetComponent<ISolutionStructureContainer>();
            var projectMark = solutionStructureContainer
                .GetProjectsByLocation(FileSystemPath.Parse(args.ProjectFilePath))
                .First();

            var project = _solution.GetProjectByMark(projectMark).NotNull();
            var targetFramework = project.TargetFrameworkIds
                // Take .NET Core first, then .NET Framework, and then .NET Standard. The comparer below will hold this order.
                .OrderBy(tfm => (!tfm.IsNetCoreApp, !tfm.IsNetFramework, !tfm.IsNetStandard))
                .First();
            var rdTargetFramework = targetFramework.ToRdTargetFrameworkInfo();

            _logger.Trace("TFM selected for project {0}: {1}", args, targetFramework);
            var assemblyInfo = project.GetOutputAssemblyInfo(targetFramework).NotNull();
            _logger.Trace("Assembly file path detected for project {0}: {1}", args, assemblyInfo.Location);
            return RdTask<RdProjectOutput>.Successful(
                new RdProjectOutput(
                    new RdTargetFrameworkIdMock(
                        rdTargetFramework.ShortName,
                        rdTargetFramework.PresentableName,
                        isNetCoreApp: rdTargetFramework.IsNetCoreApp,
                        isNetFramework: rdTargetFramework.IsNetFramework),
                    assemblyInfo.Location.ToString()));
        }
    }
}
